From cd551c05a548c8d36a6e4434db058febb8244552 Mon Sep 17 00:00:00 2001
From: Dark4MD <github.web@manu.li>
Date: Sat, 12 Jun 2021 17:38:44 +0200
Subject: [PATCH] openwrt: added TL-WR841ND/N 8M and 16M Variants to ath79

Added TL-WR841ND/N 8M and 16M v9-v11 Variants to ath79-generic (migration).
V8 is not automatically migrated as of now because of some missing migration patches and stays at ar71xx-generic.
---
 ...-WR841ND-N-Devices-for-8M-and-16M-Va.patch | 116 +++
 ...841ND-N-8M-and-16M-Variants-to-ath79.patch | 973 ++++++++++++++++++
 targets/ar71xx-generic                        |   8 +
 targets/ath79-generic                         |  38 +
 4 files changed, 1135 insertions(+)
 create mode 100644 patches/openwrt/0023-added-TP-Link-TL-WR841ND-N-Devices-for-8M-and-16M-Va.patch
 create mode 100644 patches/openwrt/0024-added-TL-WR841ND-N-8M-and-16M-Variants-to-ath79.patch

diff --git a/patches/openwrt/0023-added-TP-Link-TL-WR841ND-N-Devices-for-8M-and-16M-Va.patch b/patches/openwrt/0023-added-TP-Link-TL-WR841ND-N-Devices-for-8M-and-16M-Va.patch
new file mode 100644
index 00000000..bd01fb1f
--- /dev/null
+++ b/patches/openwrt/0023-added-TP-Link-TL-WR841ND-N-Devices-for-8M-and-16M-Va.patch
@@ -0,0 +1,116 @@
+From 2734fb40ac15f8db6eadbbd9e3438b7077876424 Mon Sep 17 00:00:00 2001
+From: Dark4MD <github.web@manu.li>
+Date: Fri, 22 Feb 2019 22:45:24 +0100
+Subject: [PATCH] added TP-Link TL-WR841ND/N Devices for 8M and 16M Variants
+
+---
+ target/linux/ar71xx/base-files/lib/ar71xx.sh |  6 +++
+ target/linux/ar71xx/image/generic-tp-link.mk | 74 ++++++++++++++++++++++++++++
+ 2 files changed, 80 insertions(+)
+ mode change 100644 => 100755 target/linux/ar71xx/image/generic-tp-link.mk
+
+diff --git a/target/linux/ar71xx/base-files/lib/ar71xx.sh b/target/linux/ar71xx/base-files/lib/ar71xx.sh
+index ccbd4e77c3..9c81ed75d0 100755
+--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh
++++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh
+@@ -255,6 +255,12 @@ tplink_board_detect() {
+ 
+ 		[ "$hwid" = '08410002' -a "$mid" = '00000002' ] && hwver=' v1.5'
+ 		;;
++	"084108"*)
++		model="TP-Link TL-WR841N/ND Mod (8M)"
++		;;
++	"084116"*)
++		model="TP-Link TL-WR841N/ND Mod (16M)"
++		;;
+ 	"084200"*)
+ 		model="TP-Link TL-WR842N/ND"
+ 		;;
+diff --git a/target/linux/ar71xx/image/generic-tp-link.mk b/target/linux/ar71xx/image/generic-tp-link.mk
+old mode 100644
+new mode 100755
+index 338b2db2d6..4e5df2c742
+--- a/target/linux/ar71xx/image/generic-tp-link.mk
++++ b/target/linux/ar71xx/image/generic-tp-link.mk
+@@ -529,3 +529,77 @@ define Device/tl-wr942n-v1
+   SUPPORTED_DEVICES := tl-wr942n-v1
+ endef
+ TARGET_DEVICES += tl-wr942n-v1
++
++define Device/tl-wr841-v8-8m
++  $(Device/tplink-8mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (8M) v8
++  BOARDNAME := TL-WR841N-v8
++  DEVICE_PROFILE := TLWR8418M
++  TPLINK_HWID := 0x08410808
++endef
++TARGET_DEVICES += tl-wr841-v8-8m
++
++define Device/tl-wr841-v9-8m
++  $(Device/tplink-8mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (8M) v9
++  BOARDNAME := TL-WR841N-v9
++  DEVICE_PROFILE := TLWR8418M
++  TPLINK_HWID := 0x08410809
++endef
++TARGET_DEVICES += tl-wr841-v9-8m
++
++define Device/tl-wr841-v10-8m
++  $(Device/tl-wr841-v9-8m)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (8M) v10
++  TPLINK_HWID := 0x08410810
++endef
++TARGET_DEVICES += tl-wr841-v10-8m
++
++define Device/tl-wr841-v11-8m
++  $(Device/tplink-8mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (8M) v11
++  BOARDNAME := TL-WR841N-v11
++  DEVICE_PROFILE := TLWR8418M
++  TPLINK_HWID := 0x08410811
++  IMAGES += factory-us.bin factory-eu.bin
++  IMAGE/factory-us.bin := append-rootfs | mktplinkfw factory -C US
++  IMAGE/factory-eu.bin := append-rootfs | mktplinkfw factory -C EU
++endef
++TARGET_DEVICES += tl-wr841-v11-8m
++
++define Device/tl-wr841-v8-16m
++  $(Device/tplink-16mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (16M) v8
++  BOARDNAME := TL-WR841N-v8
++  DEVICE_PROFILE := TLWR84116M
++  TPLINK_HWID := 0x08411608
++endef
++TARGET_DEVICES += tl-wr841-v8-16m
++
++define Device/tl-wr841-v9-16m
++  $(Device/tplink-16mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (16M) v9
++  BOARDNAME := TL-WR841N-v9
++  DEVICE_PROFILE := TLWR84116M
++  TPLINK_HWID := 0x08411609
++endef
++TARGET_DEVICES += tl-wr841-v9-16m
++
++define Device/tl-wr841-v10-16m
++  $(Device/tl-wr841-v9-16m)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (16M) v10
++  TPLINK_HWID := 0x08411610
++endef
++TARGET_DEVICES += tl-wr841-v10-16m
++
++define Device/tl-wr841-v11-16m
++  $(Device/tplink-16mlzma)
++  DEVICE_TITLE := TP-LINK TL-WR841N/ND Mod (16M) v11
++  BOARDNAME := TL-WR841N-v11
++  DEVICE_PROFILE := TLWR84116M
++  TPLINK_HWID := 0x08411611
++  IMAGES += factory-us.bin factory-eu.bin
++  IMAGE/factory-us.bin := append-rootfs | mktplinkfw factory -C US
++  IMAGE/factory-eu.bin := append-rootfs | mktplinkfw factory -C EU
++endef
++TARGET_DEVICES += tl-wr841-v11-16m
+\ No newline at end of file
+-- 
+2.11.0
+
diff --git a/patches/openwrt/0024-added-TL-WR841ND-N-8M-and-16M-Variants-to-ath79.patch b/patches/openwrt/0024-added-TL-WR841ND-N-8M-and-16M-Variants-to-ath79.patch
new file mode 100644
index 00000000..a626355b
--- /dev/null
+++ b/patches/openwrt/0024-added-TL-WR841ND-N-8M-and-16M-Variants-to-ath79.patch
@@ -0,0 +1,973 @@
+From 39d2938072680cda2cd943bcdbaafee963f0a476 Mon Sep 17 00:00:00 2001
+From: "Manu.WTF" <43337106+Dark4MD@users.noreply.github.com>
+Date: Sat, 1 Feb 2020 23:27:51 +0100
+Subject: [PATCH] added TL-WR841ND/N 8M and 16M Variants to ath79
+
+---
+ .../ath79/base-files/etc/board.d/01_leds      |   8 +
+ .../ath79/base-files/etc/board.d/02_network   |   8 +
+ .../dts/ar9341_tplink_tl-wr841-v8-16m.dts     | 165 ++++++++++++++++++
+ .../dts/ar9341_tplink_tl-wr841-v8-8m.dts      | 165 ++++++++++++++++++
+ .../dts/qca9533_tplink_tl-wr841-16m.dtsi      | 128 ++++++++++++++
+ .../ath79/dts/qca9533_tplink_tl-wr841-8m.dtsi | 128 ++++++++++++++
+ .../dts/qca9533_tplink_tl-wr841-v10-16m.dts   |  16 ++
+ .../dts/qca9533_tplink_tl-wr841-v10-8m.dts    |  16 ++
+ .../dts/qca9533_tplink_tl-wr841-v11-16m.dts   |   9 +
+ .../dts/qca9533_tplink_tl-wr841-v11-16m.dtsi  |  25 +++
+ .../dts/qca9533_tplink_tl-wr841-v11-8m.dts    |   9 +
+ .../dts/qca9533_tplink_tl-wr841-v11-8m.dtsi   |  25 +++
+ .../dts/qca9533_tplink_tl-wr841-v9-16m.dts    |  16 ++
+ .../dts/qca9533_tplink_tl-wr841-v9-8m.dts     |  16 ++
+ target/linux/ath79/image/generic-tp-link.mk   |  78 +++++++++
+ 15 files changed, 812 insertions(+)
+ create mode 100644 target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-16m.dts
+ create mode 100644 target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-8m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-16m.dtsi
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-8m.dtsi
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-16m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-8m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dtsi
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dtsi
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-16m.dts
+ create mode 100644 target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-8m.dts
+
+diff --git a/target/linux/ath79/base-files/etc/board.d/01_leds b/target/linux/ath79/base-files/etc/board.d/01_leds
+index dd0f91affa..1aa8f853e4 100755
+--- a/target/linux/ath79/base-files/etc/board.d/01_leds
++++ b/target/linux/ath79/base-files/etc/board.d/01_leds
+@@ -124,8 +124,14 @@ tplink,archer-c6-v2)
+ 	;;
+ tplink,archer-c25-v1|\
+ tplink,tl-wr841-v9|\
++tplink,tl-wr841-v9-8m|\
++tplink,tl-wr841-v9-16m|\
+ tplink,tl-wr841-v10|\
++tplink,tl-wr841-v10-8m|\
++tplink,tl-wr841-v10-16m|\
+ tplink,tl-wr841-v11|\
++tplink,tl-wr841-v11-8m|\
++tplink,tl-wr841-v11-16m|\
+ tplink,tl-wr841-v12|\
+ tplink,tl-wr842n-v3)
+ 	ucidef_set_led_netdev "wan" "WAN" "tp-link:green:wan" "eth1"
+@@ -188,6 +194,8 @@ tplink,tl-wr740n-v4|\
+ tplink,tl-wr740n-v5|\
+ tplink,tl-wr741nd-v4|\
+ tplink,tl-wr841-v8|\
++tplink,tl-wr841-v8-8m|\
++tplink,tl-wr841-v8-16m|\
+ tplink,tl-wr842n-v2)
+ 	ucidef_set_led_netdev "wan" "WAN" "tp-link:green:wan" "eth1"
+ 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:green:lan1" "switch0" "0x04"
+diff --git a/target/linux/ath79/base-files/etc/board.d/02_network b/target/linux/ath79/base-files/etc/board.d/02_network
+index 5dda551caa..f646558a0b 100755
+--- a/target/linux/ath79/base-files/etc/board.d/02_network
++++ b/target/linux/ath79/base-files/etc/board.d/02_network
+@@ -95,8 +95,14 @@ ath79_setup_interfaces()
+ 	tplink,tl-wdr3500-v1|\
+ 	tplink,tl-wr841-v7|\
+ 	tplink,tl-wr841-v9|\
++	tplink,tl-wr841-v9-8m|\
++	tplink,tl-wr841-v9-16m|\
+ 	tplink,tl-wr841-v10|\
++	tplink,tl-wr841-v10-8m|\
++	tplink,tl-wr841-v10-16m|\
+ 	tplink,tl-wr841-v11|\
++	tplink,tl-wr841-v11-8m|\
++	tplink,tl-wr841-v11-16m|\
+ 	tplink,tl-wr841-v12|\
+ 	tplink,tl-wr842n-v1|\
+ 	tplink,tl-wr842n-v3|\
+@@ -245,6 +251,8 @@ ath79_setup_interfaces()
+ 	tplink,tl-wr740n-v5|\
+ 	tplink,tl-wr741nd-v4|\
+ 	tplink,tl-wr841-v8|\
++	tplink,tl-wr841-v8-8m|\
++	tplink,tl-wr841-v8-16m|\
+ 	tplink,tl-wr842n-v2)
+ 		ucidef_set_interface_wan "eth1"
+ 		ucidef_add_switch "switch0" \
+diff --git a/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-16m.dts b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-16m.dts
+new file mode 100644
+index 0000000000..6930f51fd0
+--- /dev/null
++++ b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-16m.dts
+@@ -0,0 +1,165 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++
++#include "ar9341.dtsi"
++
++/ {
++	model = "TP-Link TL-WR841N/ND Mod (16M ATH79) v8";
++	compatible = "tplink,tl-wr841-v8-16m", "qca,ar9341";
++
++	aliases {
++		serial0 = &uart;
++		led-boot = &system;
++		led-failsafe = &system;
++		led-running = &system;
++		led-upgrade = &system;
++	};
++
++	keys {
++		compatible = "gpio-keys-polled";
++		poll-interval = <20>;
++
++		pinctrl-names = "default";
++		pinctrl-0 = <&jtag_disable_pins>;
++
++		reset {
++			label = "Reset";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++
++		rfkill {
++			label = "WiFi";
++			linux,code = <KEY_RFKILL>;
++			linux,input-type = <EV_SW>;
++			gpios = <&gpio 16 GPIO_ACTIVE_HIGH>;
++			debounce-interval = <60>;
++		};
++	};
++
++	leds {
++		compatible = "gpio-leds";
++
++		system: power {
++			label = "tp-link:green:power";
++			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
++			default-state = "on";
++		};
++
++		wlan {
++			label = "tp-link:green:wlan";
++			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
++			linux,default-trigger = "phy0tpt";
++		};
++
++		qss {
++			label = "tp-link:green:qss";
++			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
++		};
++
++		wan {
++			label = "tp-link:green:wan";
++			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
++		};
++
++		lan1 {
++			label = "tp-link:green:lan1";
++			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
++		};
++
++		lan2 {
++			label = "tp-link:green:lan2";
++			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
++		};
++
++		lan3 {
++			label = "tp-link:green:lan3";
++			gpios = <&gpio 21 GPIO_ACTIVE_LOW>;
++		};
++
++		lan4 {
++			label = "tp-link:green:lan4";
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++		};
++	};
++};
++
++&ref {
++	clock-frequency = <25000000>;
++};
++
++&uart {
++	status = "okay";
++};
++
++&gpio {
++	status = "okay";
++};
++
++&spi {
++	num-cs = <1>;
++
++	status = "okay";
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <25000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			uboot: partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x020000>;
++				read-only;
++			};
++
++			partition@20000 {
++				compatible = "tplink,firmware";
++				label = "firmware";
++				reg = <0x020000 0xfd0000>;
++			};
++
++			art: partition@ff0000 {
++				label = "art";
++				reg = <0xff0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&eth0 {
++	status = "okay";
++
++	phy-handle = <&swphy0>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++	mtd-mac-address-increment = <(-1)>;
++};
++
++&eth1 {
++	status = "okay";
++
++	mtd-mac-address = <&uboot 0x1fc00>;
++
++	pll-data = <0x06000000 0x00000101 0x00001616>;
++
++	gmac-config {
++		device = <&gmac>;
++		switch-phy-swap = <1>;
++	};
++};
++
++&wmac {
++	status = "okay";
++
++	mtd-cal-data = <&art 0x1000>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
+diff --git a/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-8m.dts b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-8m.dts
+new file mode 100644
+index 0000000000..38b4a471cb
+--- /dev/null
++++ b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8-8m.dts
+@@ -0,0 +1,165 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++
++#include "ar9341.dtsi"
++
++/ {
++	model = "TP-Link TL-WR841N/ND Mod (8M ATH79) v8";
++	compatible = "tplink,tl-wr841-v8-8m", "qca,ar9341";
++
++	aliases {
++		serial0 = &uart;
++		led-boot = &system;
++		led-failsafe = &system;
++		led-running = &system;
++		led-upgrade = &system;
++	};
++
++	keys {
++		compatible = "gpio-keys-polled";
++		poll-interval = <20>;
++
++		pinctrl-names = "default";
++		pinctrl-0 = <&jtag_disable_pins>;
++
++		reset {
++			label = "Reset";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++
++		rfkill {
++			label = "WiFi";
++			linux,code = <KEY_RFKILL>;
++			linux,input-type = <EV_SW>;
++			gpios = <&gpio 16 GPIO_ACTIVE_HIGH>;
++			debounce-interval = <60>;
++		};
++	};
++
++	leds {
++		compatible = "gpio-leds";
++
++		system: power {
++			label = "tp-link:green:power";
++			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
++			default-state = "on";
++		};
++
++		wlan {
++			label = "tp-link:green:wlan";
++			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
++			linux,default-trigger = "phy0tpt";
++		};
++
++		qss {
++			label = "tp-link:green:qss";
++			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
++		};
++
++		wan {
++			label = "tp-link:green:wan";
++			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
++		};
++
++		lan1 {
++			label = "tp-link:green:lan1";
++			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
++		};
++
++		lan2 {
++			label = "tp-link:green:lan2";
++			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
++		};
++
++		lan3 {
++			label = "tp-link:green:lan3";
++			gpios = <&gpio 21 GPIO_ACTIVE_LOW>;
++		};
++
++		lan4 {
++			label = "tp-link:green:lan4";
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++		};
++	};
++};
++
++&ref {
++	clock-frequency = <25000000>;
++};
++
++&uart {
++	status = "okay";
++};
++
++&gpio {
++	status = "okay";
++};
++
++&spi {
++	num-cs = <1>;
++
++	status = "okay";
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <25000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			uboot: partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x020000>;
++				read-only;
++			};
++
++			partition@20000 {
++				compatible = "tplink,firmware";
++				label = "firmware";
++				reg = <0x020000 0x7d0000>;
++			};
++
++			art: partition@7f0000 {
++				label = "art";
++				reg = <0x7f0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&eth0 {
++	status = "okay";
++
++	phy-handle = <&swphy0>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++	mtd-mac-address-increment = <(-1)>;
++};
++
++&eth1 {
++	status = "okay";
++
++	mtd-mac-address = <&uboot 0x1fc00>;
++
++	pll-data = <0x06000000 0x00000101 0x00001616>;
++
++	gmac-config {
++		device = <&gmac>;
++		switch-phy-swap = <1>;
++	};
++};
++
++&wmac {
++	status = "okay";
++
++	mtd-cal-data = <&art 0x1000>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-16m.dtsi b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-16m.dtsi
+new file mode 100644
+index 0000000000..ff40f961c8
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-16m.dtsi
+@@ -0,0 +1,128 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++
++#include "qca953x.dtsi"
++
++/ {
++	chosen {
++		bootargs = "console=ttyS0,115200n8";
++	};
++
++	gpio_leds: leds {
++		compatible = "gpio-leds";
++
++		wifi {
++			label = "tp-link:green:wlan";
++			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
++			linux,default-trigger = "phy0tpt";
++		};
++
++		qss_led: qss {
++			label = "tp-link:green:qss";
++			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
++		};
++
++		wan {
++			label = "tp-link:green:wan";
++			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
++		};
++
++		lan1 {
++			label = "tp-link:green:lan1";
++			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
++		};
++
++		lan2 {
++			label = "tp-link:green:lan2";
++			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
++		};
++
++		lan3 {
++			label = "tp-link:green:lan3";
++			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
++		};
++
++		lan4 {
++			label = "tp-link:green:lan4";
++			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	keys {
++		compatible = "gpio-keys";
++
++		reset {
++			label = "Reset button";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++
++		rfkill {
++			label = "RFKILL button";
++			linux,code = <KEY_RFKILL>;
++			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++	};
++};
++
++&uart {
++	status = "okay";
++};
++
++&spi {
++	status = "okay";
++	num-cs = <1>;
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <25000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			uboot:	partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x020000>;
++				read-only;
++			};
++
++			partition@20000 {
++				compatible = "tplink,firmware";
++				label = "firmware";
++				reg = <0x020000 0xfd0000>;
++			};
++
++			art: partition@ff0000 {
++				label = "art";
++				reg = <0xff0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&eth1 {
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
++
++&eth0 {
++	status = "okay";
++	phy-handle = <&swphy4>;
++
++	mtd-mac-address = <&uboot 0x1fc00>;
++	mtd-mac-address-increment = <1>;
++};
++
++&wmac {
++	status = "okay";
++	mtd-cal-data = <&art 0x1000>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-8m.dtsi b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-8m.dtsi
+new file mode 100644
+index 0000000000..2097a299d2
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-8m.dtsi
+@@ -0,0 +1,128 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++
++#include "qca953x.dtsi"
++
++/ {
++	chosen {
++		bootargs = "console=ttyS0,115200n8";
++	};
++
++	gpio_leds: leds {
++		compatible = "gpio-leds";
++
++		wifi {
++			label = "tp-link:green:wlan";
++			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
++			linux,default-trigger = "phy0tpt";
++		};
++
++		qss_led: qss {
++			label = "tp-link:green:qss";
++			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
++		};
++
++		wan {
++			label = "tp-link:green:wan";
++			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
++		};
++
++		lan1 {
++			label = "tp-link:green:lan1";
++			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
++		};
++
++		lan2 {
++			label = "tp-link:green:lan2";
++			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
++		};
++
++		lan3 {
++			label = "tp-link:green:lan3";
++			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
++		};
++
++		lan4 {
++			label = "tp-link:green:lan4";
++			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	keys {
++		compatible = "gpio-keys";
++
++		reset {
++			label = "Reset button";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++
++		rfkill {
++			label = "RFKILL button";
++			linux,code = <KEY_RFKILL>;
++			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
++			debounce-interval = <60>;
++		};
++	};
++};
++
++&uart {
++	status = "okay";
++};
++
++&spi {
++	status = "okay";
++	num-cs = <1>;
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <25000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			uboot:	partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x020000>;
++				read-only;
++			};
++
++			partition@20000 {
++				compatible = "tplink,firmware";
++				label = "firmware";
++				reg = <0x020000 0x7d0000>;
++			};
++
++			art: partition@7f0000 {
++				label = "art";
++				reg = <0x7f0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&eth1 {
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
++
++&eth0 {
++	status = "okay";
++	phy-handle = <&swphy4>;
++
++	mtd-mac-address = <&uboot 0x1fc00>;
++	mtd-mac-address-increment = <1>;
++};
++
++&wmac {
++	status = "okay";
++	mtd-cal-data = <&art 0x1000>;
++	mtd-mac-address = <&uboot 0x1fc00>;
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-16m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-16m.dts
+new file mode 100644
+index 0000000000..0d177ca926
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-16m.dts
+@@ -0,0 +1,16 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-16m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v10-16m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (16M ATH79) v10";
++
++	aliases {
++		led-boot = &qss_led;
++		led-failsafe = &qss_led;
++		led-running = &qss_led;
++		led-upgrade = &qss_led;
++	};
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-8m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-8m.dts
+new file mode 100644
+index 0000000000..c50b3f2168
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v10-8m.dts
+@@ -0,0 +1,16 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-8m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v10-8m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (8M ATH79) v10";
++
++	aliases {
++		led-boot = &qss_led;
++		led-failsafe = &qss_led;
++		led-running = &qss_led;
++		led-upgrade = &qss_led;
++	};
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dts
+new file mode 100644
+index 0000000000..02de2c94ac
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dts
+@@ -0,0 +1,9 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-v11-16m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v11-16m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (16M ATH79) v11";
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dtsi b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dtsi
+new file mode 100644
+index 0000000000..c5e72c6930
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-16m.dtsi
+@@ -0,0 +1,25 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-16m.dtsi"
++
++/ {
++	aliases {
++		led-boot = &system_led;
++		led-failsafe = &system_led;
++		led-running = &system_led;
++		led-upgrade = &system_led;
++	};
++};
++
++&gpio_leds {
++	system_led: system {
++		label = "tp-link:green:system";
++		gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
++	};
++
++	wan_orange {
++		label = "tp-link:orange:wan";
++		gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
++	};
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dts
+new file mode 100644
+index 0000000000..d73cde55ba
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dts
+@@ -0,0 +1,9 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-v11-8m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v11-8m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (8M ATH79) v11";
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dtsi b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dtsi
+new file mode 100644
+index 0000000000..a1663abee2
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11-8m.dtsi
+@@ -0,0 +1,25 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-8m.dtsi"
++
++/ {
++	aliases {
++		led-boot = &system_led;
++		led-failsafe = &system_led;
++		led-running = &system_led;
++		led-upgrade = &system_led;
++	};
++};
++
++&gpio_leds {
++	system_led: system {
++		label = "tp-link:green:system";
++		gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
++	};
++
++	wan_orange {
++		label = "tp-link:orange:wan";
++		gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
++	};
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-16m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-16m.dts
+new file mode 100644
+index 0000000000..a68c5ddc7c
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-16m.dts
+@@ -0,0 +1,16 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-16m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v9-16m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (16M ATH79) v9";
++
++	aliases {
++		led-boot = &qss_led;
++		led-failsafe = &qss_led;
++		led-running = &qss_led;
++		led-upgrade = &qss_led;
++	};
++};
+diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-8m.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-8m.dts
+new file mode 100644
+index 0000000000..dbecc8a7d7
+--- /dev/null
++++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v9-8m.dts
+@@ -0,0 +1,16 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++/dts-v1/;
++
++#include "qca9533_tplink_tl-wr841-8m.dtsi"
++
++/ {
++	compatible = "tplink,tl-wr841-v9-8m", "qca,qca9533";
++	model = "TP-Link TL-WR841N/ND Mod (8M ATH79) v9";
++
++	aliases {
++		led-boot = &qss_led;
++		led-failsafe = &qss_led;
++		led-running = &qss_led;
++		led-upgrade = &qss_led;
++	};
++};
+diff --git a/target/linux/ath79/image/generic-tp-link.mk b/target/linux/ath79/image/generic-tp-link.mk
+index 2ae84bd2ed..6b8468b32e 100644
+--- a/target/linux/ath79/image/generic-tp-link.mk
++++ b/target/linux/ath79/image/generic-tp-link.mk
+@@ -333,6 +333,84 @@ define Device/tplink_tl-wr710n-v2.1
+ endef
+ TARGET_DEVICES += tplink_tl-wr710n-v2.1
+ 
++define Device/tplink_tl-wr841-v8-8m
++  $(Device/tplink-8mlzma)
++  ATH_SOC := ar9341
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (8M ATH79) v8
++  TPLINK_HWID := 0x08410808
++  SUPPORTED_DEVICES += tl-wr841n-v8 tl-wr841n-v8-8m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v8-8m
++
++define Device/tplink_tl-wr841-v9-8m
++  $(Device/tplink-8mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (8M ATH79) v9
++  TPLINK_HWID := 0x08410809
++  SUPPORTED_DEVICES += tl-wr841n-v9 tl-wr841n-v9-8m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v9-8m
++
++define Device/tplink_tl-wr841-v10-8m
++  $(Device/tplink-8mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (8M ATH79) v10
++  TPLINK_HWID := 0x08410810
++  SUPPORTED_DEVICES += tl-wr841n-v9 tl-wr841n-v9-8m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v10-8m
++
++define Device/tplink_tl-wr841-v11-8m
++  $(Device/tplink-8mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (8M ATH79) v11
++  TPLINK_HWID := 0x08410811
++  SUPPORTED_DEVICES += tl-wr841n-v11 tl-wr841n-v11-8m
++  IMAGES += factory-us.bin factory-eu.bin
++  IMAGE/factory-us.bin := append-rootfs | mktplinkfw factory -C US
++  IMAGE/factory-eu.bin := append-rootfs | mktplinkfw factory -C EU
++endef
++TARGET_DEVICES += tplink_tl-wr841-v11-8m
++
++define Device/tplink_tl-wr841-v8-16m
++  $(Device/tplink-16mlzma)
++  ATH_SOC := ar9341
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (16M ATH79) v8
++  TPLINK_HWID := 0x08411608
++  SUPPORTED_DEVICES += tl-wr841n-v8 tl-wr841n-v8-16m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v8-16m
++
++define Device/tplink_tl-wr841-v9-16m
++  $(Device/tplink-16mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (16M ATH79) v9
++  TPLINK_HWID := 0x08411609
++  SUPPORTED_DEVICES += tl-wr841n-v9 tl-wr841n-v9-16m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v9-16m
++
++define Device/tplink_tl-wr841-v10-16m
++  $(Device/tplink-16mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (16M ATH79) v10
++  TPLINK_HWID := 0x08411610
++  SUPPORTED_DEVICES += tl-wr841n-v9 tl-wr841n-v9-16m
++endef
++TARGET_DEVICES += tplink_tl-wr841-v10-16m
++
++define Device/tplink_tl-wr841-v11-16m
++  $(Device/tplink-16mlzma)
++  ATH_SOC := qca9533
++  DEVICE_TITLE := TP-Link TL-WR841N/ND Mod (16M ATH79) v11
++  TPLINK_HWID := 0x08411611
++  SUPPORTED_DEVICES += tl-wr841n-v11 tl-wr841n-v11-8m
++  IMAGES += factory-us.bin factory-eu.bin
++  IMAGE/factory-us.bin := append-rootfs | mktplinkfw factory -C US
++  IMAGE/factory-eu.bin := append-rootfs | mktplinkfw factory -C EU
++endef
++TARGET_DEVICES += tplink_tl-wr841-v11-16m
++
+ define Device/tplink_tl-wr842n-v1
+   $(Device/tplink-8m)
+   ATH_SOC := ar7241
+-- 
+2.29.2.windows.2
+
diff --git a/targets/ar71xx-generic b/targets/ar71xx-generic
index df7dbd97..f9939017 100644
--- a/targets/ar71xx-generic
+++ b/targets/ar71xx-generic
@@ -398,6 +398,14 @@ device('tp-link-re450', 're450-v1', {
 	class = 'tiny', -- Only 6M of usable Firmware space
 })
 
+device('tp-link-tl-wr841n-nd-mod-8m-v8', 'tl-wr841-v8-8m', {
+        factory = false,
+})
+
+device('tp-link-tl-wr841n-nd-mod-16m-v8', 'tl-wr841-v8-16m', {
+        factory = false,
+})
+
 
 -- Ubiquiti
 
diff --git a/targets/ath79-generic b/targets/ath79-generic
index 5537a7ea..32003304 100644
--- a/targets/ath79-generic
+++ b/targets/ath79-generic
@@ -97,3 +97,41 @@ device('tp-link-archer-d50-v1', 'tplink_archer-d50-v1', {
 	factory = false,
 	broken = true, -- 64M ath9k + ath10k & power LED not working
 })
+
+device('tp-link-tl-wr841n-nd-mod-8m-ath79-v8', 'tplink_tl-wr841-v8-8m', {
+        factory = false,
+})
+
+device('tp-link-tl-wr841n-nd-mod-8m-ath79-v9', 'tplink_tl-wr841-v9-8m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-8m-v9'},
+})
+
+device('tp-link-tl-wr841n-nd-mod-8m-ath79-v10', 'tplink_tl-wr841-v10-8m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-8m-v10'},
+})
+
+device('tp-link-tl-wr841n-nd-mod-8m-ath79-v11', 'tplink_tl-wr841-v11-8m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-8m-v11'},
+})
+
+device('tp-link-tl-wr841n-nd-mod-16m-ath79-v8', 'tplink_tl-wr841-v8-16m', {
+        factory = false,
+})
+
+device('tp-link-tl-wr841n-nd-mod-16m-ath79-v9', 'tplink_tl-wr841-v9-16m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-16m-v9'},
+})
+
+device('tp-link-tl-wr841n-nd-mod-16m-ath79-v10', 'tplink_tl-wr841-v10-16m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-16m-v10'},
+})
+
+device('tp-link-tl-wr841n-nd-mod-16m-ath79-v11', 'tplink_tl-wr841-v11-16m', {
+        factory = false,
+        aliases = {'tp-link-tl-wr841n-nd-mod-16m-v11'},
+})
-- 
2.29.2.windows.2

